internal/database/sqlcommon/identity_sql.go

```diff
@@ -0,0 +1,250 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sqlcommon
+
+import (
+	"context"
+	"database/sql"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/hyperledger/firefly/internal/i18n"
+	"github.com/hyperledger/firefly/internal/log"
+	"github.com/hyperledger/firefly/pkg/database"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+)
+
+var (
+	identityColumns = []string{
+		"id",
+		"did",
+		"parent",
+		"itype",
+		"namespace",
+		"name",
+		"description",
+		"profile",
+		"messages_claim",
+		"messages_verification",
+		"messages_update",
+		"created",
+		"updated",
+	}
+	identityFilterFieldMap = map[string]string{
+		"identity":              "identity_id",
+		"type":                  "itype",
+		"messages.claim":        "messages_claim",
+		"messages.verification": "messages_verification",
+		"messages.update":       "messages_update",
+	}
+)
+
+func (s *SQLCommon) attemptIdentityUpdate(ctx context.Context, tx *txWrapper, identity *fftypes.Identity) (int64, error) {
+	identity.Updated = fftypes.Now()
+	return s.updateTx(ctx, tx,
+		sq.Update("identities").
+			Set("did", identity.DID).
+			Set("parent", identity.Parent).
+			Set("itype", identity.Type).
+			Set("namespace", identity.Namespace).
+			Set("name", identity.Name).
+			Set("description", identity.Description).
+			Set("profile", identity.Profile).
+			Set("messages_claim", identity.Messages.Claim).
+			Set("messages_verification", identity.Messages.Verification).
+			Set("messages_update", identity.Messages.Update).
+			Set("updated", identity.Updated).
+			Where(sq.Eq{
+				"id": identity.ID,
+			}),
+		func() {
+			s.callbacks.UUIDCollectionNSEvent(database.CollectionIdentities, fftypes.ChangeEventTypeUpdated, identity.Namespace, identity.ID)
+		})
+}
+
+func (s *SQLCommon) attemptIdentityInsert(ctx context.Context, tx *txWrapper, identity *fftypes.Identity) (err error) {
+	identity.Created = fftypes.Now()
+	identity.Updated = identity.Created
+	_, err = s.insertTx(ctx, tx,
+		sq.Insert("identities").
+			Columns(identityColumns...).
+			Values(
+				identity.ID,
+				identity.DID,
+				identity.Parent,
+				identity.Type,
+				identity.Namespace,
+				identity.Name,
+				identity.Description,
+				identity.Profile,
+				identity.Messages.Claim,
+				identity.Messages.Verification,
+				identity.Messages.Update,
+				identity.Created,
+				identity.Updated,
+			),
+		func() {
+			s.callbacks.UUIDCollectionNSEvent(database.CollectionIdentities, fftypes.ChangeEventTypeCreated, identity.Namespace, identity.ID)
+		})
+	return err
+}
+
+func (s *SQLCommon) UpsertIdentity(ctx context.Context, identity *fftypes.Identity, optimization database.UpsertOptimization) (err error) {
```

Comment:
I went straight to the optimization enabled upsert, as it seems like a good pattern.
I hope we'll roll this out universally across the code over time.